drivers/sensor/sensirion/sen6x/sen6x_decoder.c

```diff
@@ -0,0 +1,461 @@
+				    size_t *out_divisor)
+{
+	bool is_signed;
+	size_t divisor;
+
+	switch (chan) {
+	case SENSOR_CHAN_PM_1_0:
+	case SENSOR_CHAN_PM_2_5:
+	case SENSOR_CHAN_PM_4_0:
+	case SENSOR_CHAN_PM_10:
+	case SENSOR_CHAN_PM_0_5_NUM:
+	case SENSOR_CHAN_PM_1_0_NUM:
+	case SENSOR_CHAN_PM_2_5_NUM:
+	case SENSOR_CHAN_PM_4_0_NUM:
+	case SENSOR_CHAN_PM_10_NUM:
+		is_signed = false;
+		divisor = 10;
+		break;
+	case SENSOR_CHAN_AMBIENT_HUMIDITY:
+		is_signed = true;
+		divisor = 100;
+		break;
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		is_signed = true;
+		divisor = 200;
+		break;
+	case SENSOR_CHAN_VOC_INDEX:
+		is_signed = true;
+		divisor = 10;
+		break;
+	case SENSOR_CHAN_NOX_INDEX:
+		is_signed = true;
+		divisor = 10;
+		break;
+	case SENSOR_CHAN_CO2:
+		is_signed = false;
+		divisor = 1;
+		break;
+	case SENSOR_CHAN_HCHO:
+		is_signed = false;
+		divisor = 10;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	if (out_is_signed) {
+		*out_is_signed = is_signed;
+	}
+	if (out_divisor) {
+		*out_divisor = divisor;
+	}
+	return 0;
+}
+
+int sen6x_encode(const struct device *dev, const struct sensor_chan_spec *const channels,
+		 const size_t num_channels, uint8_t *buf)
+{
+	const struct sen6x_config *cfg = dev->config;
+	struct sen6x_encoded_data *edata = (struct sen6x_encoded_data *)buf;
+	uint64_t cycles;
+	int err;
+	bool needs_measured_values = false;
+	bool needs_separate_number_concentration = false;
+
+	if (cfg->model != SEN6X_MODEL_SEN60) {
+		for (size_t i = 0; i < num_channels; i++) {
+			if (sen6x_get_channel_index(edata, channels[i].chan_type) >=
+			    MAX_MEASURED_VALUES_COUNT) {
+				needs_separate_number_concentration = true;
+			} else {
+				needs_measured_values = true;
+			}
+		}
+	}
+
+	err = sensor_clock_get_cycles(&cycles);
+	if (err != 0) {
+		return err;
+	}
+
+	edata->header = (struct sen6x_encoded_header){
+		.model = cfg->model,
+		.timestamp = sensor_clock_cycles_to_ns(cycles),
+		.has_data_ready = needs_measured_values || needs_separate_number_concentration,
+		.has_measured_values = needs_measured_values,
+		.has_number_concentration = needs_separate_number_concentration,
+	};
+
+	return 0;
+}
+
+static int sen6x_decoder_get_frame_count(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
+					 uint16_t *frame_count)
+{
+	struct sen6x_encoded_data *edata = (struct sen6x_encoded_data *)buffer;
+	int ret;
+	uint8_t *channel_data;
+	uint16_t value;
+	int16_t value_signed;
+	bool is_signed;
+
+	if (chan_spec.chan_idx != 0) {
+		return -ENOTSUP;
+	}
+
+	ret = sen6x_get_channel_index(edata, chan_spec.chan_type);
+	if (ret < 0) {
+		return ret;
+	}
+	if (ret < MAX_MEASURED_VALUES_COUNT && !edata->header.has_measured_values) {
+		return -ENODATA;
+	}
+	if (ret >= MAX_MEASURED_VALUES_COUNT && !edata->header.has_number_concentration) {
+		return -ENODATA;
+	}
+
+	ret = sen6x_get_channel_format(chan_spec.chan_type, &is_signed, NULL);
+	if (ret) {
+		return ret;
+	}
+	channel_data = &edata->channels[ret * 3];
+
```

Comment:
seems to be a more logical grouping. also, what is 3 here?

```suggestion

	channel_data = &edata->channels[ret * 3];
```